aspnet/fundamentals/routing.rst

```diff
@@ -1,10 +1,283 @@
+
+.. literalinclude:: routing/sample/RoutingSample/Startup.cs
+  :language: c#
+  :lines: 27-49
+  :dedent: 12
+  :emphasize-lines: 9,16
+
+.. _link-generation:
+
+Link Generation
+---------------
+
+Routing is also used to generate URLs based on route definitions. This is used by helpers to generate links to known actions on MVC :doc:`controllers </mvc/controllers/index>`, but can also be used independent of MVC. Given a set of route values, and optionally a route name, you can produce a ``VirtualPathContext`` object. Using the ``VirtualPathContext`` object along with a ``RouteCollection``, you can generate a ``VirtualPath``. ``IRouter`` implementations participate in link generation through the ``GetVirtualPath`` method.
+
+.. tip:: Learn more about `UrlHelper <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Routing/UrlHelper/index.html?highlight=urlhelper>`_ and :doc:`Routing to Controller Actions </mvc/controllers/routing>`.
+
+The example below shows how to generate a link to a route given a dictionary of route values and a ``RouteCollection``.
+
+.. literalinclude:: routing/sample/RoutingSample/Startup.cs
+  :language: c#
+  :lines: 45-69
+  :dedent: 8
+  :emphasize-lines: 9,14,17-18,23
+
+The ``VirtualPath`` generated at the end of the sample above is ``/package/create/123``.
+
+The second parameter to the ``VirtualPathContext`` constructor is a collection of `ambient values`. Ambient values provide convenience by limiting the number of values a developer must specify within a certain request context. The current route values of the current request are considered ambient values for link generation. For example, in an MVC application if you are in the About action of the HomeController, you don't need to specify the controller route value to link to the Index action (the ambient value of Home will be used). 
+
+Ambient values that don't match a parameter are ignored, and ambient values are also ignored when an explicitly-provided value overrides it, going from left to right in the URL.
+
+Values that are explicitly provided but which don't match anything are added to the query string.
+
+.. list-table:: Generating Links
+	:header-rows: 1
+
+	* - Matched Route
+	  - Ambient Values
+	  - Explicit Values
+	  - Result
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home"
+	  - action="About"
+	  - ``/Home/About``
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home"
+	  - controller="Order",action="About"
+	  - ``/Order/About``
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home",color="Red"
+	  - action="About"
+	  - ``/Home/About``
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home"
+	  - action="About",color="Red"
+	  - ``/Home/About?color=Red``
+
+If a route parameter has a default value that doesn't match a parameter and that value is explicitly provided, it must match the default value. For example:
```

Comment:
`If a route parameter ...` -> `If a route ...` 

This is for the case where there **isn't** a corresponding parameter.